### ISSUE TITLE
Error: “Unable to find a basis set…” with CFOUR

### ISSUE THREAD

**Issue by tobirolinew** [psi4/psi4#2608]
Hi,

I just started to use PSI4 and wanted to test whether my installation is able to work with CFOUR. When running PSI4 with the following input:

```

    #!CFOUR test

    molecule {
    O
    H 1 R
    H 1 R 2 A

    R=0.958
    A=104.5
    }

    set {
    cfour_CALC_level=CCSD(T)
    cfour_BASIS=AUG-PVQZ
    cfour_SCF_CONV=12
    cfour_CC_CONV=12
    }

    energy(‘cfour’)
```

Psi4 is complaining about the basis set:

    QcdbException BasisSetNotFound: BasisSet::construct: Unable to find a basis set for atom 1 for key BASIS among:
    Shell Entries: [‘O’]
    Basis Sets: [(’(AUTO)’, ‘(AUTO)’, None)]
    File Path: /scr/tobias/psi.1379622…cfour.14c28cc5, /scr/tobias/psi.1379622…cfour.14c28cc5, /xstorage/tobias/bin/psi4/share/psi4/basis
    Input Blocks:

    Traceback (most recent call last):
    File “/xstorage/tobias/bin/psi4/bin/psi4”, line 338, in
    exec(content)
    File “”, line 31, in
    File “/xstorage/tobias/bin/psi4/lib//python3.8/site-packages/psi4/driver/driver.py”, line 501, in energy
    wfn = procedures[‘energy’][lowername](lowername, molecule=molecule, **kwargs)
    File “/xstorage/tobias/bin/psi4/lib//python3.8/site-packages/psi4/driver/procrouting/interface_cfour.py”, line 185, in run_cfour
    cfour_infile.write(write_zmat(lowername, dertype, molecule))
    File “/xstorage/tobias/bin/psi4/lib//python3.8/site-packages/psi4/driver/procrouting/interface_cfour.py”, line 461, in write_zmat
    qbs = core.BasisSet.build(molecule, “BASIS”, core.get_global_option(‘BASIS’))
    File “/xstorage/tobias/bin/psi4/lib//python3.8/site-packages/psi4/driver/p4util/python_helpers.py”, line 85, in _pybuild_basis
    bs, basisdict = qcdb.BasisSet.pyconstruct(mol.to_dict(),
    File “/xstorage/tobias/bin/psi4/lib//python3.8/site-packages/psi4/driver/qcdb/libmintsbasisset.py”, line 696, in pyconstruct
    bs, msg, ecp = BasisSet.construct(parser, mol,
    File “/xstorage/tobias/bin/psi4/lib//python3.8/site-packages/psi4/driver/qcdb/libmintsbasisset.py”, line 918, in construct
    raise BasisSetNotFound(f’BasisSet::construct: Unable to find a basis set for atom {at + 1} for key {key} among:\n{text2}’)

    psi4.driver.qcdb.exceptions.BasisSetNotFound: BasisSet::construct: Unable to find a basis set for atom 1 for key BASIS among:
    Shell Entries: [‘O’]
    Basis Sets: [(’(AUTO)’, ‘(AUTO)’, None)]
    File Path: /scr/tobias/psi.1379622…cfour.14c28cc5, /scr/tobias/psi.1379622…cfour.14c28cc5, /xstorage/tobias/bin/psi4/share/psi4/basis
    Input Blocks:

    Printing out the relevant lines from the Psithon → Python processed input file:
    core.IO.set_default_namespace("")
    core.set_global_option(“CFOUR_CALC_LEVEL”, “CCSD(T)”)
    core.set_global_option(“CFOUR_BASIS”, “AUG-PVQZ”)
    core.set_global_option(“CFOUR_SCF_CONV”, 12)
    core.set_global_option(“CFOUR_CC_CONV”, 12)
    → energy(‘cfour’)

    !----------------------------------------------------------------------------------!
    ! !
    ! BasisSet::construct: Unable to find a basis set for atom 1 for key BASIS among: !
    ! Shell Entries: [‘O’] !
    ! Basis Sets: [(’(AUTO)’, ‘(AUTO)’, None)] !
    ! File Path: /scr/tobias/psi.1379622…cfour.14c28cc5, !
    ! /scr/tobias/psi.1379622…cfour.14c28cc5, !
    ! /xstorage/tobias/bin/psi4/share/psi4/basis !
    ! Input Blocks: !
    ! !
    !----------------------------------------------------------------------------------!

Could you please help me what causes the problem? Note that `psi4 --test` also fails for CFOUR:

`lib/python3.8/site-packages/psi4/tests/test_addons.py::test_cfour FAILED`

Thank you in advance for your help.

**Comment by JonathonMisiewicz** [comment]
Please let us know how you installed Psi4. In particular, we need to know the version number that should appear in the header of your input file.

@loriab, I hate to keep pinging you on things, but I feel that the CFOUR interface is your department.

**Comment by tobirolinew** [comment]
Dean Jonathon,

Thank you for your quick reply. I installed Psi4 by executing the bash script `Psi4conda-1.6-py38-Linux-x86_64.sh` (with the command `bash Psi4conda-1.6-py38-Linux-x86_64.sh`). I guess I installed the 1.6 version:

```
Psi4: An Open-Source Ab Initio Electronic Structure Package
                               Psi4 1.6 release
```
Thanks again.

Best regards,

Roland

**Comment by tobirolinew** [comment]
Thanks Jonathon. If it is simpler, I can give you access to our cluster (by sending you the login data via email) where you can test the problematic input.

**Comment by loriab** [comment]
Thanks for trying out Psi4 and the Psi4/Cfour interface, @tobirolinew. I can probably fix your immediate problem, but please read on. :-)

There's three ways of running a Cfour calc through a more flexible front-end right now:
* QCEngine (already present in a Psi4 installation. takes a dictionary as input. geometry must be in Cartesians.)
* QCDB (extra install. depends on QCEngine. inputs look much like Psi4/Cfour only `psi4.` becomes `qcdb.`. can use Z-Matrices.)
* Psi4/Cfour (the first in time of these three. what you're trying now.)

I'm trying to deprecate this last one in favor of the better maintained first ones. Your job would look something like the below through QCEngine (code below not tested).

```
import psi4
import qcengine as qcng

h2o = psi4.geometry("""
    O
    H 1 R
    H 1 R 2 A

    R=0.958
    A=104.5
""")

atomicinput = {
    "driver": "energy",
    "molecule": h2o.to_schema(dtype=2),
    "model": {
        "method": "ccsd(t)",
        "basis": "aug-pvqz",
    },
    "keywords": {
        "SCF_CONV": 12,
        "CC_CONV": 12,
    },
}

atomicresult = qcng.compute(atomicinput, "cfour")

import pprint
pprint.pprint(atomic_result.dict())
```

What types of calcs are you ultimately looking to run? How would you like to proceed: the QCEngine or Psi4/Cfour route? Thanks!

**Comment by tobirolinew** [comment]
Thanks @loriab for your detailed answer. In fact, I would like to do focal-point analyses for small molecules (including HF, MP2, CC, DBOC, and relativistic corrections) in a somewhat automated form. Psi4, which is able to execute MRCC and CFOUR, seems to me an optimal frame for this purpose (I saw that certain FPA schemes are already implemented in it). 

The Psi4/Cfour route looks simpler for me. Basically I will use Cartesian coordinates, but in some cases it would be useful to do CCSD(T) optimizations via CFOUR (requiring internal coordinates).

How to execute the solution you posted with Psi4? I tried to put it into a file `test.dat` and run `psi4 test.dat`, but -- after correcting the possible typo [`atomic_result.dict()` instead of `atomicresult.dict()`], I received the following error: 

```
{'error': {'error_message': 'QCEngine Execution Error:\n'
                            'Traceback (most recent call last):\n'
                            '  File '
                            '"/xstorage/tobias/bin/psi4/lib//python3.8/site-packages/qcengine/util.py", '
                            'line 114, in compute_wrapper\n'
                            '    yield metadata\n'
                            '  File '
                            '"/xstorage/tobias/bin/psi4/lib//python3.8/site-packages/qcengine/compute.py", '
                            'line 91, in compute\n'
                            '    output_data = executor.compute(input_data, '
                            'config)\n'
                            '  File '
                            '"/xstorage/tobias/bin/psi4/lib//python3.8/site-packages/qcengine/programs/cfour/runner.py", '
                            'line 71, in compute\n'
                            '    job_inputs = self.build_input(input_model, '
                            'config)\n'
                            '  File '
                            '"/xstorage/tobias/bin/psi4/lib//python3.8/site-packages/qcengine/programs/cfour/runner.py", '
                            'line 137, in build_input\n'
                            '    cfourrec["infiles"]["GENBAS"] = '
                            'genbas.read_text()\n'
                            '  File '
                            '"/xstorage/tobias/bin/psi4/lib/python3.8/pathlib.py", '
                            'line 1236, in read_text\n'
                            "    with self.open(mode='r', encoding=encoding, "
                            'errors=errors) as f:\n'
                            '  File '
                            '"/xstorage/tobias/bin/psi4/lib/python3.8/pathlib.py", '
                            'line 1222, in open\n'
                            '    return io.open(self, mode, buffering, '
                            'encoding, errors, newline,\n'
                            '  File '
                            '"/xstorage/tobias/bin/psi4/lib/python3.8/pathlib.py", '
                            'line 1078, in _opener\n'
                            '    return self._accessor.open(self, flags, '
                            'mode)\n'
                            'FileNotFoundError: [Errno 2] No such file or '
                            'directory: '
                            "'/xstorage/tobias/bin/cfour/basis/GENBAS'\n",
           'error_type': 'unknown_error',
           'extras': None},
 'extras': None,
 'id': None,
 'input_data': {'driver': 'energy',
                'keywords': {'CC_CONV': 12, 'SCF_CONV': 12},
                'model': {'basis': 'aug-pvqz', 'method': 'ccsd(t)'},
                'molecule': {'atom_labels': ['', '', ''],
                             'atomic_numbers': [8, 1, 1],
                             'fix_com': False,
                             'fix_orientation': False,
                             'fragment_charges': [0.0],
                             'fragment_multiplicities': [1],
                             'fragments': [[0, 1, 2]],
                             'geometry': [0.0,
                                          0.0,
                                          -0.12403886030029525,
                                          0.0,
                                          -1.431430901356359,
                                          0.9842933627189459,
                                          0.0,
                                          1.431430901356359,
                                          0.9842933627189459],
                             'mass_numbers': [16, 1, 1],
                             'masses': [15.99491461957,
                                        1.00782503223,
                                        1.00782503223],
                             'molecular_charge': 0.0,
                             'molecular_multiplicity': 1,
                             'name': 'H2O',
                             'provenance': {'creator': 'QCElemental',
                                            'routine': 'qcelemental.molparse.from_string',
                                            'version': 'v0.24.0'},
                             'real': [True, True, True],
                             'schema_name': 'qcschema_molecule',
                             'schema_version': 2,
                             'symbols': ['O', 'H', 'H'],
                             'validated': True},
                'provenance': {'cpu': 'Intel(R) Xeon(R) CPU E7-4870 v2 @ '
                                      '2.30GHz',
                               'creator': 'QCEngine',
                               'hostname': 'nod10',
                               'qcengine_version': 'v0.23.0',
                               'username': 'tobias',
                               'version': 'v0.23.0',
                               'wall_time': 1.2955999374389648}},
 'success': False}
```
 
Could you please help me what causes this error?

**Comment by loriab** [comment]
I can answer more fully later, but check where the `GENBAS` file lives in your CFOUR installation. It looks like it's looking for it at `/xstorage/tobias/bin/cfour/basis/GENBAS`. (This could be the trouble in the Psi4/Cfour route, too.) I don't remember the exact fallback procedure for GENBAS, but if it's missing from that spot, you could copy it to there or copy the one from the psi4 install (approx <install>/share/psi4/basis/GENBAS).

**Comment by tobirolinew** [comment]
@loriab Thanks, the GENBAS file was indeed missing from directory `/xstorage/tobias/bin/cfour/basis/`. After copying one into that directory, the input you provided now works on our cluster. I look forward to getting to know the Psi4/CFOUR route.

**Comment by tobirolinew** [comment]
@loriab, I noticed that you posted a fix for my Psi4/Cfour problem at https://github.com/psi4/psi4/pull/2615. Thank you for that! As I understand correctly, you have the updated files using which all Cfour test jobs run successfully. If so, could you please help me which files have been modified, and from where I can download them?

**Comment by loriab** [comment]
@tobirolinew, pretty much all you need for runtime are the changes in this file, https://github.com/psi4/psi4/pull/2615/files#diff-3ca8e1e030771d01af412b9603a6f8e06d4c04e44dbcb40a4f7248ff8120efdd . So, you could either make those edits in your local install, or compile psi4 from master, or grab the latest conda package next week after I build them again. I'll also probably back-port these fixes into v1.6.1.
